### PR TITLE
Refactor RegExp builtin

### PR DESCRIPTION
**Pull request by dbatyai**
JerryScript-DCO-1.0-Signed-off-by: Dániel Bátyai dbatyai@inf.u-szeged.hu

### CODE REVIEW

**Comment by dbatyai** [comment]
Benchmark | Heap (bytes) | Perf (sec) |
----: | ----: | ----: | 
regexp-dna.js | 1924160 -> 1312304: +31.798% | 30.950 -> 2.680: +91.341% |
string-tagcloud.js | 1347128 -> 1345008: +0.157% | 28.973 -> 3.340: +88.472% |
string-unpack-code.js | 856912 -> 752064: +12.235% | 63.103 -> 9.697: +84.634% |
v8/regexp.js | 2276152 -> 2154480: +5.345% | 257.516 -> 164.916: +35.959% |


Binary size:

Binary (bytes) | master | patch | Diff |
----: | ----: | ----: | ----: |
.text | 118460 | 117588 | -872 bytes |

**Comment by renatahodovan** [comment]
###### JerryScript revision
0392f04a


###### Build platform
Linux-4.15.0-62-generic-x86_64-with-Ubuntu-18.04-bionic

###### Build steps
```
./tools/build.py --clean --debug --compile-flag=-fsanitize=address \
--compile-flag=-m32 --compile-flag=-fno-omit-frame-pointer \
--compile-flag=-fno-common --compile-flag=-g \
--strip=off --system-allocator=on --logging=on \
--linker-flag=-fuse-ld=gold --error-messages=on --profile=es2015-subset
```

###### Test case
```javascript
RegExp().compile(this.RegExp.prototype)
```


###### Output
```text
ICE: Assertion 'target_bc_p != NULL' failed at jerryscript/jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.c(ecma_builtin_regexp_prototype_compile):214.
Error: ERR_FAILED_INTERNAL_ASSERTION
```

###### Backtrace
```text
bt
#0  0xf7fd5079 in __kernel_vsyscall ()
#1  0xf77fc832 in raise () from /lib/i386-linux-gnu/libc.so.6
#2  0xf77fdcc1 in abort () from /lib/i386-linux-gnu/libc.so.6
#3  0x56579773 in jerry_port_fatal (code=ERR_FAILED_INTERNAL_ASSERTION) at jerryscript/jerry-port/default/default-fatal.c:71
#4  0x56611f77 in jerry_fatal (code=ERR_FAILED_INTERNAL_ASSERTION) at jerryscript/jerry-core/jrt/jrt-fatals.c:58
#5  0x56611fb8 in jerry_assert_fail (assertion=0x566a7640 "target_bc_p != NULL", file=0x566a7580 "jerryscript/jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.c", function=0x5668f6e0 <__func__.4225.lto_priv.290> "ecma_builtin_regexp_prototype_compile", line=214) at jerryscript/jerry-core/jrt/jrt-fatals.c:82
#6  0x565ccdfe in ecma_builtin_regexp_prototype_compile (this_arg=4126148883, pattern_arg=4126148931, flags_arg=72) at jerryscript/jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.c:214
#7  0x565cbfa9 in ecma_builtin_regexp_prototype_dispatch_routine (builtin_routine_id=76, this_arg_value=4126148883, arguments_list=0xffffc3d0, arguments_number=1) at jerryscript/jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.inc.h:77
#8  0x56624af9 in ecma_builtin_dispatch_routine (builtin_object_id=ECMA_BUILTIN_ID_REGEXP_PROTOTYPE, builtin_routine_id=76, this_arg_value=4126148883, arguments_list_p=0xffffc3d0, arguments_list_len=1) at jerryscript/jerry-core/ecma/builtin-objects/ecma-builtins.c:1021
#9  0x56624d5a in ecma_builtin_dispatch_call (obj_p=0xf5f008b0, this_arg_value=4126148883, arguments_list_p=0xffffc66c, arguments_list_len=1) at jerryscript/jerry-core/ecma/builtin-objects/ecma-builtins.c:1046
#10 0x56638597 in ecma_op_function_call (func_obj_p=0xf5f008b0, this_arg_value=4126148883, arguments_list_p=0xffffc66c, arguments_list_len=1) at jerryscript/jerry-core/ecma/operations/ecma-function-object.c:729
#11 0x565e5ca6 in opfunc_call (frame_ctx_p=0xffffc6e0) at jerryscript/jerry-core/vm/vm.c:581
#12 0x565f707d in vm_execute (frame_ctx_p=0xffffc6e0, arg_p=0x0, arg_list_len=0) at jerryscript/jerry-core/vm/vm.c:3636
#13 0x565f7981 in vm_run (bytecode_header_p=0xf5100f50, this_binding_value=4126149459, lex_env_p=0xf5d007b0, parse_opts=0, arg_list_p=0x0, arg_list_len=0) at jerryscript/jerry-core/vm/vm.c:3756
#14 0x565e4f3d in vm_run_global (bytecode_p=0xf5100f50) at jerryscript/jerry-core/vm/vm.c:282
#15 0x566463bc in jerry_run (func_val=4126149027) at jerryscript/jerry-core/api/jerry.c:576
#16 0x56642bbb in main (argc=3, argv=0xffffcb14) at jerryscript/jerry-main/main-unix.c:743
```

<sup>Found by [Fuzzinator](http://fuzzinator.readthedocs.io/) with [grammarinator](https://github.com/renatahodovan/grammarinator). </sup>

**Comment by dbatyai** [comment]
I've made some additional changes, here are the current results:

Benchmark | Perf (sec) |
----: | ----: | 
regexp-dna.js |  30.912 -> 2.372: +92.328% |
string-tagcloud.js | 29.107 -> 3.312: +88.622% |
string-unpack-code.js | 63.057 -> 9.758: +84.525% |
v8/regexp.js | 257.516 -> 64.217: +75.062% |


Binary size:

Binary (bytes) | master | patch | Diff |
----: | ----: | ----: | ----: |
.text | 118460 | 117588 | -584 bytes |

File: jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.c [hunk 1]
```diff
@@ -45,6 +45,121 @@
  * @{
  */
 
+#if ENABLED (JERRY_ES2015)
+static ecma_value_t
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.c]: Missing function description.

File: jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.c [hunk 2]
```diff
@@ -45,6 +45,121 @@
...
+  if (bc_p != NULL)
+  {
+    *flags_p = bc_p->header.status_flags;
+  }
+
+  return ECMA_VALUE_EMPTY;
+} /* ecma_builtin_regexp_prototype_flags_helper */
+
+static ecma_value_t
+ecma_builtin_regexp_prototype_get_flags (ecma_value_t this_arg) /**< this argument */
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.c]: Ditto

File: jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.c [hunk 3]
```diff
@@ -45,6 +45,121 @@
...
+  if (flags & RE_FLAG_MULTILINE)
+  {
+    ecma_stringbuilder_append_byte (&result, LIT_CHAR_LOWERCASE_M);
+  }
+
+  return ecma_make_string_value (ecma_stringbuilder_finalize (&result));
+} /* ecma_builtin_regexp_prototype_get_flags */
+
+static ecma_value_t
+ecma_builtin_regexp_prototype_get_source (ecma_value_t this_arg) /**< this argument */
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.c]: Ditto.

File: jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.c [hunk 4]
```diff
@@ -45,6 +45,121 @@
...
+  {
+    ecma_ref_ecma_string (ecma_get_string_from_value (bc_p->source));
+    return bc_p->source;
+  }
+
+  return ecma_make_string_value (ecma_get_magic_string (LIT_MAGIC_STRING_EMPTY_NON_CAPTURE_GROUP));
+} /* ecma_builtin_regexp_prototype_get_source */
+
+static ecma_value_t
+ecma_builtin_regexp_prototype_get_global (ecma_value_t this_arg) /**< this argument */
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.c]: Ditto.

File: jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.c [hunk 5]
```diff
@@ -45,6 +45,121 @@
...
+  if (ECMA_IS_VALUE_ERROR (ret_value))
+  {
+    return ret_value;
+  }
+
+  return ecma_make_boolean_value (flags & RE_FLAG_GLOBAL);
+} /* ecma_builtin_regexp_prototype_get_global */
+
+static ecma_value_t
+ecma_builtin_regexp_prototype_get_ignorecase (ecma_value_t this_arg) /**< this argument */
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.c]: Ditto.

File: jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.c [hunk 6]
```diff
@@ -45,6 +45,121 @@
...
+  if (ECMA_IS_VALUE_ERROR (ret_value))
+  {
+    return ret_value;
+  }
+
+  return ecma_make_boolean_value (flags & RE_FLAG_IGNORE_CASE);
+} /* ecma_builtin_regexp_prototype_get_ignorecase */
+
+static ecma_value_t
+ecma_builtin_regexp_prototype_get_multiline (ecma_value_t this_arg) /**< this argument */
```
Comment (rerobika) [inline jerry-core/ecma/builtin-objects/ecma-builtin-regexp-prototype.c]: Ditto.

**Pull request merged**